server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/SmtpMtPriorityMessageHookTest.java

```diff
@@ -0,0 +1,147 @@
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.smtpserver;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.james.smtpserver.SMTPServerTestSystem.BOB;
+import static org.apache.james.smtpserver.SMTPServerTestSystem.PASSWORD;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.Base64;
+
+import org.apache.commons.net.smtp.SMTPClient;
+import org.apache.mailet.AttributeName;
+import org.apache.mailet.Mail;
+import org.assertj.core.api.SoftAssertions;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class SmtpMtPriorityMessageHookTest {
+
+    private final SMTPServerTestSystem testSystem = new SMTPServerTestSystem();
+    private static final AttributeName MAIL_PRIORITY_ATTRIBUTE_NAME = AttributeName.of("MAIL_PRIORITY");
+
+    @BeforeEach
+    void setUp() throws Exception {
+        testSystem.setUp("smtpserver-mtPriority.xml");
+    }
+
+    @AfterEach
+    void tearDown() {
+        testSystem.smtpServer.destroy();
+    }
+
+    @Test
+    void ehloShouldAdvertiseMtPriorityExtension() throws Exception {
+        SMTPClient smtpProtocol = new SMTPClient();
+        InetSocketAddress bindedAddress = testSystem.getBindedAddress();
+        smtpProtocol.connect(bindedAddress.getAddress().getHostAddress(), bindedAddress.getPort());
+        authenticate(smtpProtocol);
+        smtpProtocol.sendCommand("EHLO localhost");
+
+        SoftAssertions.assertSoftly(softly -> {
+            softly.assertThat(smtpProtocol.getReplyCode()).isEqualTo(250);
+            softly.assertThat(smtpProtocol.getReplyString()).contains("250 MT-PRIORITY");
+        });
+    }
+
+    @ParameterizedTest
+    @CsvSource({"-9,0", "-8,0", "-7,0", "-6,0", "-5,0", "-4,0", "-3,0", "-2,0", "-1,0", "0,0", "1,1",
+        "2,2", "3,3", "4,4", "5,5", "6,6", "7,7", "8,8", "9,9"})
```

Comment:
Yes true actually the mail queue implementations do not support negative priorities as of today yet I believe this is more of a mistake and mis-understanding of legacy implementors rather than a concious choice.

It would be great if we could modify the JMS mailque to support this and alter PriorityMailQueueContract accordingly, and thus having here:

```suggestion
    @CsvSource({"-9,-9", "-8,-8", "-7,-7", "-6,-6", "-5,-5", "-4,-4", "-3,-3", "-2,-2", "-1,-1", "0,0", "1,1",
        "2,2", "3,3", "4,4", "5,5", "6,6", "7,7", "8,8", "9,9"})
```

While we do not violate the spec (that accounts for such rounding) we could at least plan this as an improvement.

Reply:
If all goes well, I plan to implement this improvement in a separate PR

Reply:
> Yes true actually the mail queue implementations do not support negative priorities as of today yet I believe this is more of a mistake and mis-understanding of legacy implementors rather than a concious choice.
> 
> It would be great if we could modify the JMS mailque to support this and alter PriorityMailQueueContract accordingly, and thus having here:
> 
> While we do not violate the spec (that accounts for such rounding) we could at least plan this as an improvement.

I am currently looking for an optimal solution, considering that queues implement the The Jakarta Messaging API.
The Jakarta Messaging API defines ten levels of priority value, with 0 as the lowest priority and 9 as the highest.  Any ideas on how to use negative priorities? What might a modification or an alternative of the JMS look like?